crates/data_components/src/github.rs

```diff
@@ -0,0 +1,275 @@
+            path_builder.append_value(&node.path);
+            size_builder.append_value(node.size.unwrap_or(0));
+            sha_builder.append_value(&node.sha);
+            mode_builder.append_value(&node.mode);
+            url_builder.append_value(&node.url);
+        }
+
+        let record_batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(name_builder.finish()),
+                Arc::new(path_builder.finish()),
+                Arc::new(size_builder.finish()),
+                Arc::new(sha_builder.finish()),
+                Arc::new(mode_builder.finish()),
+                Arc::new(url_builder.finish()),
+            ],
+        )
+        .context(UnableToConstructRecordBatchSnafu)?;
+
+        Ok(vec![record_batch])
+    }
+
+    async fn fetch_git_tree(
+        &self,
+        owner: &str,
+        repo: &str,
+        tree_sha: &str,
+    ) -> Result<GitTree, Box<dyn std::error::Error + Send + Sync>> {
+        let endpoint = format!(
+            "https://api.github.com/repos/{owner}/{repo}/git/trees/{tree_sha}?recursive=true"
+        );
+
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(SPICE_USER_AGENT));
+        headers.insert(
+            ACCEPT,
+            HeaderValue::from_static("application/vnd.github.v3+json"),
+        );
+
+        if let Ok(header) = HeaderValue::from_str(&format!("token {}", self.token)) {
+            headers.insert(AUTHORIZATION, header);
+        }
+
+        tracing::debug!("fetch_git_tree: endpoint: {}", endpoint);
+
+        let response = self.client.get(&endpoint).headers(headers).send().await?;
+
+        if response.status().is_success() {
+            let git_tree = response.json::<GitTree>().await?;
+            tracing::trace!("fetch_git_tree returned {} entities", git_tree.tree.len());
+            return Ok(git_tree);
+        }
+
+        #[allow(clippy::single_match_else)]
+        match response.status().as_u16() {
+            404 => {
+                let err_msg = format!(
+                    "Github API ({endpoint}) failed with status code {}; Is org `{owner}`, repo `{repo}` and git tree `{tree_sha}` correct?",
+                    response.status()
+                );
+                Err(err_msg.into())
+            }
+            _ => {
+                let err_msg = format!(
+                    "Github API ({endpoint}) failed with status code {}",
+                    response.status()
+                );
+                Err(err_msg.into())
+            }
+        }
+    }
+}
+
+fn extract_name_from_path(path: &str) -> Option<&str> {
+    Path::new(path).file_name().and_then(|name| name.to_str())
+}
+
+#[derive(Debug, Deserialize)]
+struct GitTree {
+    tree: Vec<GitTreeNode>,
+}
+
+#[derive(Debug, Deserialize)]
+struct GitTreeNode {
+    path: String,
+    mode: String,
+    #[serde(rename = "type")]
+    node_type: String,
```

Comment:
I think you can do `r#node_type: String,` if you like? I don't know what's default, or what we prefer in the repo.

Reply:
Didn't know that you can do this, thanks! Tried but found out that it is not convenient to use this field after that. Keeping as-is for now.